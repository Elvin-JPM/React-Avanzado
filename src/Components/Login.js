import { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { authLogin, authLoginFailure, uiResetError } from "../store/actions";
import { getUI } from "../store/selectors";

function Login({ handleShowMessage }) {
  const dispatch = useDispatch();
  const { isFetching, error } = useSelector(getUI);
  // const [isFetching, setIsFetching] = useState(false);
  // const [error, setError] = useState(null);
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [remember, setRemember] = useState(null);
  const resetError = () => {
    dispatch(uiResetError());
  };

  const requestBody = {
    email,
    password,
  };

  const handleChange = (inputId, newValue) => {
    if (inputId === "email") setEmail(newValue);
    if (inputId === "password") setPassword(newValue);
  };

  const handleRememberUser = (event) => {
    setRemember(event.target.checked);
  };

  const handleClick = async () => {
    if (email && password) {
      try {
        await dispatch(authLogin(requestBody, remember));
      } catch (error) {
        dispatch(authLoginFailure(error));
        handleShowMessage(error.message, "showFailure");
        setTimeout(() => {
          handleShowMessage("", "doNotShow");
        }, 2000);
      }
    } else {
      handleShowMessage("Email and password required", "showFailure");
      setTimeout(() => {
        handleShowMessage("", "doNotShow");
      }, 2000);
    }
  };

  return (
    <>
      <form>
        <Input
          id="email"
          type="email"
          value={email}
          placeholder="email"
          handleChange={handleChange}
        ></Input>
        <Input
          id="password"
          type="password"
          value={password}
          placeholder="password"
          handleChange={handleChange}
        ></Input>

        <input
          role="button"
          type="button"
          value={"Login"}
          onClick={handleClick}
        ></input>
        <div>
          <input
            type="checkbox"
            id="remember"
            value="remember"
            onChange={handleRememberUser}
          ></input>
          <label htmlFor="remember">Remember username and password</label>
        </div>

        <label>Don't have an account?</label>
        <Link to="/signup">Sign up here</Link>
        {error && <div onClick={resetError}>{error.message}</div>}
      </form>
    </>
  );
}

function Input({ id, type, value, placeholder, handleChange }) {
  return (
    <input
      type={type}
      value={value}
      placeholder={placeholder}
      onChange={(event) => {
        handleChange(id, event.target.value);
      }}
    ></input>
  );
}

export default Login;
