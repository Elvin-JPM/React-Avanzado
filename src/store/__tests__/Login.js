import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import Login from "../../Components/Login";

describe("Login", () => {
  const state = { ui: { isFetching: false, error: null } };
  const store = {
    getState: () => state,
    subscribe: () => {},
    dispatch: () => {},
  };

  const history = createMemoryHistory();

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
  test("snapshot", () => {
    const renderComp = renderComponent();
    expect(renderComp.container).toMatchSnapshot();
  });

  test("should dispatch outhLogin action", () => {
    renderComponent();
    const usernameInput = screen.getByLabelText(/username/);
    const password = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole("button");
  });
});
