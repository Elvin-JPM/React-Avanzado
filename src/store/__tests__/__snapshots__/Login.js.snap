// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot 1`] = `
<div>
  <form>
    <input
      placeholder="email"
      type="email"
      value=""
    />
    <input
      placeholder="password"
      type="password"
      value=""
    />
    <input
      type="button"
      value="Login"
    />
    <div>
      <input
        id="remember"
        type="checkbox"
        value="remember"
      />
      <label
        for="remember"
      >
        Remember username and password
      </label>
    </div>
    <label>
      Don't have an account?
    </label>
    <a
      href="/signup"
    >
      Sign up here
    </a>
  </form>
</div>
`;
